cachi2/core/models/sbom.py

```diff
@@ -93,6 +106,235 @@ def _unique_components(cls, components: list[Component]) -> list[Component]:
+):
+    """SPDX Package External Reference for category package-manager and type purl.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: Literal["PACKAGE-MANAGER"]
+    referenceType: Literal["purl"]
+
+
+class SPDXPackageExternalRefSecurityPURL(
+    SPDXPackageExternalRefSecurity, SPDXPackageExternalRefReferenceLocatorURI
+):
+    """SPDX Package External Reference for category package-manager and type purl.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: Literal["SECURITY"]
+    referenceType: Literal["cpe23Type"]
+
+
+SPDXPackageExternalRefPackageManagerType = Annotated[
+    SPDXPackageExternalRefPackageManagerPURL,
+    pydantic.Field(discriminator="referenceType"),
+]
+
+SPDXPackageExternalRefSecurityType = Annotated[
+    SPDXPackageExternalRefSecurityPURL,
+    pydantic.Field(discriminator="referenceType"),
+]
+
+
+SPDXPackageExternalRefType = Annotated[
+    Union[SPDXPackageExternalRefPackageManagerType, SPDXPackageExternalRefSecurityType],
+    pydantic.Field(discriminator="referenceCategory"),
+]
+
+
+class SPDXPackageAnnotation(pydantic.BaseModel):
+    """SPDX Package Annotation.
+
+    Compliant to the SPDX specification:
+    https://github.com/spdx/spdx-spec/blob/development/v2.3/schemas/spdx-schema.json#L237
+    """
+
+    model_config = pydantic.ConfigDict(frozen=True)
+
+    annotator: str
+    annotationDate: str
+    annotationType: Literal["OTHER", "REVIEW"]
+    comment: str
+
+    def __hash__(self) -> int:
+        return hash((self.annotator, self.annotationDate, self.annotationType, self.comment))
+
+
+def _extract_purls(from_refs: list[SPDXPackageExternalRefType]) -> list[str]:
+    return [ref.referenceLocator for ref in from_refs if ref.referenceType == "purl"]
+
+
+def _parse_purls(purls: list[str]) -> list[PackageURL]:
+    return [PackageURL.from_string(purl) for purl in purls if purl]
+
+
+class SPDXPackage(pydantic.BaseModel):
+    """SPDX Package.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/
+    """
+
+    SPDXID: Optional[str] = None
+    name: str
+    versionInfo: Optional[str] = None
+    externalRefs: list[SPDXPackageExternalRefType] = []
+    annotations: list[SPDXPackageAnnotation] = []
+    downloadLocation: str = "NOASSERTION"
+
+    def __lt__(self, other: "SPDXPackage") -> bool:
+        return (self.SPDXID or "") < (other.SPDXID or "")
+
+    def __hash__(self) -> int:
+        return hash(
+            hash(self.SPDXID)
+            + hash(self.name)
+            + hash(self.versionInfo)
+            + hash(self.downloadLocation)
+            + sum(hash(e) for e in self.externalRefs)
+            + sum(hash(a) for a in self.annotations)
+        )
```

Comment:
I'm curious, why isn't this model frozen if you want to make it hashable? Same with 
`SPDXRelation`